Allow Tabs to be controlled

matthewbryant95 opened the pull request:
## Symptom

It's currently not clear from the interface how to control the active tab from a higher component. 

## Problem

The interface for controlling the active `Tab` isn't natural. 

## Solution

1. Add an `activeTabIndex` prop to the `Tabs` component which allows a higher component to control the active tab when used in conjunction with the pre-existing `onTabChange` function (which we now ensure is called whenever another `Tab` is clicked). 
2. Deprecate the `active` property on the `Tab` component.

* Also, add a `defaultActiveTabIndex` prop to `Tabs`, which allows the initially active tab to be set whilst yielding control of the active `Tab` to `Tabs`.

## Verification

* Added a 'controlled Tabs' section to the storybook.
* Added unit tests for the new `defaultActiveTabIndex` and `activeTabIndex` props.
* Added unit tests for the `onTabChange` prop on the `Tabs` component.

matthewbryant95 commented:
Anyone know why `ci/circleci: visual-test` is failing... looks like a credentials issue, do I need to set something up?

wmui51 commented on packages/tabs/src/Tabs.js:
```diff
@@ -75,6 +78,10 @@ class Tabs extends Component {
      * When variant is set to "canvas", the effective alignment will always be "left"
      */
     align: PropTypes.oneOf(AVAILABLE_ALIGNMENTS),
+    /**
```
The component already allows for allowing the passing of a css `className` through using `...props` so we do not need to have it as an explicit prop here.

matthewbryant95 replied:
I didn't add this prop, it is already being used, I just added it to the `propTypes` for increased type checking and documentation.

wmui51 replied:
Got it, but please still remove this as we do not don't usually to type checking for `className` and we it is in the `README` that users can apply `className` for styling.

wmui51 commented:
@matthewbryant95 you do not have to worry about the visual-tests running for now. That is a permissions thing and I will trigger them for you. I can provide you with credentials to run them locally if you want, but you would only need this if there are any visual changes, which does not seem to be the case with this PR.

matthewbryant95 commented:
> @matthewbryant95 you do not have to worry about the visual-tests running for now. That is a permissions thing and I will trigger them for you. I can provide you with credentials to run them locally if you want, but you would only need this if there are any visual changes, which does not seem to be the case with this PR.

Great, thanks!

matthewbryant95 commented:
Any chance I could get a full review of this, my team need to integrate the change into our project as soon as possible? Thanks!

wmui51 commented:
@matthewbryant95 we will take a look and test this out this week for a full code review. Thanks again for contributing this back.

garyps commented on packages/tabs/src/Tab.js:
```diff
@@ -6,6 +6,9 @@ import TabPresenter from "./presenters/TabPresenter";
     /**
+     * DEPRECATED: Use the activeTab or defaultActiveTabIndex prop on the Tabs
+     * component instead.
+     *
```
Thought we can just remove `active` from propTypes/defaultProps for `Tab`, just like the other props (`onMouseLeave`,`showDivider`, etc. ) that not open to end users, but will be overwritten via payloads from parent's `renderTabs` function. And just deconstruct it from `...otherProps` when needed in the render function?

wmui51 replied:
I think we can leave `active` for now to let the users know that it is on the road to being deprecated.  For the first line,  `DEPRECATED: Use the activeTab or defaultActiveTabIndex prop on the Tabs...` the actual prop is `activeTabIndex` and not `activeTab`?

matthewbryant95 commented:
Thanks @garyps! As it stands, the behaviour for anyone using the `active` prop should not change because of this PR. My thinking was that I would leave the prop documented and working as before but with a deprecation warning, then perhaps set a release for its removal. So for example, release this change as `v1.2.0`, then I can add the release for removal to the docs with the deprecation warning, e.g. "DEPRECATED: Use the activeTab or defaultActiveTabIndex prop on the Tabs component instead. Set for removal in v2.0.0" )?

Also happy to remove `active` from `propTypes`/`defaultProps` and then push this as a breaking change though if that's what you think is best, not sure what process you normally follow for removing parts of the API. Let me know what you think and I'll make changes as necessary 😄

garyps commented:
> Thanks @garyps! As it stands, the behaviour for anyone using the `active` prop should not change because of this PR. 
> 

I think in your current solution, directly setting `active` prop in children `Tab` component will not continue working? The default active index will not read from this prop (but from the new props), and it was also removed when children changed (`getDerivedStateFromProps`). So if we release this PR, and after users upgrade their packages, their code will not work (default active tab will always be the first one, even they pass active prop in 2nd/3rd/... tab).

matthewbryant95 commented:
Ah, sorry, you're right! I have pushed a fix 👍 When it was changed it was not working as expected anyway since `getDerivedStateFromProps` on `Tabs` does not get called when a prop on `Tab` is changed, so this does not change the behaviour, I don't think?

Again, happy to undo those changes if you'd prefer to push as a breaking change anyway, just let me know 😃

wmui51 commented:
@matthewbryant95 we definitely do not want a `BREAKING CHANGE` for this and will push one a few releases down the line after folks get used to using the `activeTab` prop instead.

For your last `fix` commit, I think you may want to change that to a `refactor` as a `fix` commit will include the message into the `CHANGELOG` and in the current release that is not an issue that needs to be fixed but was introduced in this pr and may be a bit confusing to have in the `CHANGELOG`.

garyps commented:
Thanks @matthewbryant95 , I saw that the latest commit do fix the issue for initial active state on using old `active` prop. 

> Ah, sorry, you're right! I have pushed a fix 👍 When it was changed it was not working as expected anyway since `getDerivedStateFromProps` on `Tabs` does not get called when a prop on `Tab` is changed, so this does not change the behaviour, I don't think?

For `getDerivedStateFromProps`, hmm, yes, React cannot monitor a deep change on a child prop, and I am ok with your solutions. Just curious if user also changes/provides a `key` on one child `Tab`, or if user pushes/pops one child in/out the children `Tab`'s array, thought the `getDerivedStateFromProps` may will be called?

matthewbryant95 commented:
On thinking about it again, I think `getDerivedStateFromProps` is going to get called, because if you are controlling the from a parent component, then its re-render on state change will cause its children to re-render. I've reverted that part of the change and updated the PR description to reflect that.

wmui51 commented:
@garyps have you had a chance to look at the last refactor?

wmui51 commented:
@matthewbryant95 sorry for the delay, I took a look today and ran through it on `storybook` and the code looks good and it looks like it works as expected. I had a minor edit on the code side in a comment.

I also do think that we should remove the `storybook` example for the "controlled tabs." The main reason is that when you look at the story itself it does not have an example on how it is implemented which could lead to more questions.

<img width="1327" alt="Screen Shot 2020-03-03 at 3 22 03 PM" src="https://user-images.githubusercontent.com/11857921/75829555-5f493480-5d63-11ea-9d76-e633c9e5f8a2.png">

Since there is no example of the code shown in `storybook` the example mainly shows that it is possible to have controlled tabs which I think is already implied with your documentation on the props table.

wmui51 commented:
@matthewbryant95 thanks again for this and sorry for the delay. I have merged your contribution in.
